crypto-intrinsics/src/x86_64.rs

```diff
@@ -0,0 +1,148 @@
+///
+/// Performs an unsigned addition of the destination operand (first operand),
+/// the source operand (second operand) and the overflow-flag (OF) and stores
+/// the result in the destination operand. The destination operand is a
+/// general-purpose register, whereas the source operand can be a
+/// general-purpose register or memory location. The state of OF represents a
+/// carry from a previous addition. The instruction sets the OF flag with the
+/// carry generated by the unsigned addition of the operands.
+///
+/// The ADOX instruction is executed in the context of multi-precision
+/// addition, where we add a series of operands with a carry-chain. At the
+/// beginning of a chain of additions, we execute an instruction to zero the OF
+/// (e.g. XOR).
+///
+/// # Safety
+///
+/// This function requires support for the Intel ADX (Multi-Precision Add-Carry
+/// Instruction) extension to the x86 instruction set.
+#[inline(always)]
+pub unsafe fn adox(a: &mut u64, b: u64) {
+    asm! {
+        "adox {0}, {1}",
+        inout(reg) *a,
+        in(reg) b
+    };
+}
+
+/// Move if zero.
+///
+/// Uses a `test` instruction to check if the given `condition` value is
+/// equal to zero, then calls `cmovz` (a.k.a. `cmove`) to conditionally move
+/// `src` to `dst` when `condition` is equal to zero.
+#[inline(always)]
+pub fn cmovz(condition: u64, src: u64, dst: &mut u64) {
+    unsafe {
+        asm! {
+            "test {0}, {0}",
+            "cmovz {1}, {2}",
+            in(reg) condition,
+            inlateout(reg) *dst,
+            in(reg) src
+        };
+    }
+}
+
+/// Move if not zero.
+///
+/// Uses a `test` instruction to check if the given `condition` value is not
+/// equal to zero, then calls `cmovnz` (a.k.a. `cmovne`) to conditionally move
+/// `src` to `dst` when `condition` is nonzero.
+#[inline(always)]
+pub fn cmovnz(condition: u64, src: u64, dst: &mut u64) {
+    unsafe {
+        asm! {
+            "test {0}, {0}",
+            "cmovnz {1}, {2}",
+            in(reg) condition,
+            inlateout(reg) *dst,
+            in(reg) src
+        };
+    }
+}
+
+/// Unsigned multiply without affecting flags.
+///
+/// Performs an unsigned multiplication of the implicit source operand and the
+/// specified source operand (the third operand) and stores the low half of the
+/// result in the second destination (second operand), the high half of the
+/// result in the first destination operand (first operand), without reading or
+/// writing the arithmetic flags.
+///
+/// This enables efficient programming where the software can interleave add
+/// with carry operations and multiplications.
+///
+/// If the first and second operand are identical, it will contain the high
+/// half of the multiplication result.
+///
+/// # Safety
+///
+/// This function requires support for the Intel BMI2 (Bit Manipulation
+/// Instruction Set 2) extension to the x86 instruction set.
+#[target_feature(enable = "bmi2")]
```

Comment:
I was able to add this for MULX, as it requires BMI2.

I tried to do the same thing with ADX, but unfortunately:

```
error[E0658]: the target feature `adx` is currently unstable
  --> crypto-intrinsics/src/x86_64.rs:26:18
   |
26 | #[target_feature(enable = "adx")]
   |                  ^^^^^^^^^^^^^^
```

Reply:
It looks like a PR to stabilize ADX stalled some time in 2019: https://github.com/rust-lang/rust/pull/60109